### PR TITLE
httptypes: WriteTo to return error

### PR DESCRIPTION
**Pull request by gyuho**
This is a new pull request following up with https://github.com/coreos/etcd/pull/3574.
@xiang90 suggested just to return error. This PR changes it to such interface.
This API change should be compatible in general, since the code before this did
not return any value to be evaluated. I might be missing some changes, and
wasn't sure about what would be the best way to handle this error.

And this would make it seem inconsistent with other APIs as in `rafthttp`[1]
and `error`[2]:

```
// TODO(xiangli): handle http write errors
func (e HTTPError) WriteTo(w http.ResponseWriter) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(e.Code)
	b, err := json.Marshal(e)
	if err != nil {
		plog.Panicf("marshal HTTPError should never fail (%v)", err)
	}
	w.Write(b)
}

type writerToResponse interface {
	WriteTo(w http.ResponseWriter)
}

func (e Error) WriteTo(w http.ResponseWriter) {
	w.Header().Add("X-Etcd-Index", fmt.Sprint(e.Index))
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(e.statusCode())
	fmt.Fprintln(w, e.toJsonString())
}

```

Please let me know if you have any feedback. And thanks!

---
1. https://github.com/coreos/etcd/blob/master/rafthttp/http.go#L63-L65
2. https://github.com/coreos/etcd/blob/master/error/error.go#L154-L159

### CODE REVIEW

File: etcdserver/etcdhttp/capability.go [hunk 1]
```diff
@@ -84,14 +84,17 @@ func isCapabilityEnabled(c capability) bool {
...
 			return
 		}
 		fn(w, r)
 	}
 }
 
-func notCapable(w http.ResponseWriter, c capability) {
+func notCapable(w http.ResponseWriter, r *http.Request, c capability) {
 	herr := httptypes.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Not capable of accessing %s feature during rolling upgrades.", c))
 	herr.WriteTo(w)
```
Comment (xiang90) [inline etcdserver/etcdhttp/capability.go]: remove line 96

File: etcdserver/etcdhttp/httptypes/errors_test.go [hunk 2]
```diff
@@ -24,7 +24,9 @@ import (
 func TestHTTPErrorWriteTo(t *testing.T) {
 	err := NewHTTPError(http.StatusBadRequest, "what a bad request you made!")
 	rr := httptest.NewRecorder()
-	err.WriteTo(rr)
+	if e := err.WriteTo(rr); e != nil {
+		t.Errorf("marshal HTTPError should never fail (%v)", e)
```
Comment (xiang90) [inline etcdserver/etcdhttp/httptypes/errors_test.go]: i do not understand this.

**Comment by gyuho** [comment]
@xiang90 just pushed another commit to your comments. Thanks for review!

File: etcdserver/etcdhttp/httptypes/errors_test.go [hunk 3]
```diff
@@ -24,7 +24,9 @@ import (
 func TestHTTPErrorWriteTo(t *testing.T) {
 	err := NewHTTPError(http.StatusBadRequest, "what a bad request you made!")
 	rr := httptest.NewRecorder()
-	err.WriteTo(rr)
+	if e := err.WriteTo(rr); e != nil {
+		t.Errorf("HTTPError.WriteTo error (%v)", e)
```
Comment (xiang90) [inline etcdserver/etcdhttp/httptypes/errors_test.go]: use t.Fatalf(). if this failed, we should not continue the test.

**Comment by xiang90** [comment]
defer to @yichengq

File: etcdserver/etcdhttp/capability.go [hunk 4]
```diff
@@ -84,14 +84,16 @@ func isCapabilityEnabled(c capability) bool {
...
 		fn(w, r)
 	}
 }
 
-func notCapable(w http.ResponseWriter, c capability) {
+func notCapable(w http.ResponseWriter, r *http.Request, c capability) {
 	herr := httptypes.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Not capable of accessing %s feature during rolling upgrades.", c))
-	herr.WriteTo(w)
+	if err := herr.WriteTo(w); err != nil {
+		plog.Debugf("errors writing HTTPError (%v) to %s", err, r.RemoteAddr)
```
Comment (yichengq) [inline etcdserver/etcdhttp/capability.go]: s/errors/error

File: etcdserver/etcdhttp/client.go [hunk 5]
```diff
@@ -706,7 +706,7 @@ func getID(p string, w http.ResponseWriter) (types.ID, bool) {
 	}
 	id, err := types.IDFromString(idStr)
 	if err != nil {
-		writeError(w, httptypes.NewHTTPError(http.StatusNotFound, fmt.Sprintf("No such member: %s", idStr)))
+		writeError(w, nil, httptypes.NewHTTPError(http.StatusNotFound, fmt.Sprintf("No such member: %s", idStr)))
```
Comment (yichengq) [inline etcdserver/etcdhttp/client.go]: We cannot use nil here, or it may panic.

File: etcdserver/etcdhttp/http.go [hunk 6]
```diff
@@ -49,10 +49,14 @@ func writeError(w http.ResponseWriter, err error) {
 	case *etcdErr.Error:
 		e.WriteTo(w)
```
Comment (yichengq) [inline etcdserver/etcdhttp/http.go]: why not check error here?

File: etcdserver/etcdhttp/httptypes/errors.go [hunk 7]
```diff
@@ -35,15 +35,17 @@ func (e HTTPError) Error() string {
...
 
-// TODO(xiangli): handle http write errors
-func (e HTTPError) WriteTo(w http.ResponseWriter) {
+func (e HTTPError) WriteTo(w http.ResponseWriter) error {
 	w.Header().Set("Content-Type", "application/json")
 	w.WriteHeader(e.Code)
 	b, err := json.Marshal(e)
 	if err != nil {
-		plog.Panicf("marshal HTTPError should never fail (%v)", err)
+		return err
```
Comment (yichengq) [inline etcdserver/etcdhttp/httptypes/errors.go]: I think it should use panic code. At any case, marshal should never fail.

File: etcdserver/etcdhttp/client.go (hunk 5 above)
Comment (gyuho) [inline etcdserver/etcdhttp/client.go]: Just fixed https://github.com/gyuho/etcd/commit/ea1d0da300b5d88caac02a3a70d75de5accec023

File: etcdserver/etcdhttp/httptypes/errors.go (hunk 7 above)
Comment (gyuho) [inline etcdserver/etcdhttp/httptypes/errors.go]: @xiang90  Should I change it to `Panicf` in the logs? I have `Debugf`s now.

Thanks for review.

Comment (xiang90) [inline etcdserver/etcdhttp/httptypes/errors.go]: Let's keep the original line44, the log.Panicf one.

**Comment by xiang90** [comment]
LGTM

File: etcdserver/etcdhttp/httptypes/errors.go [hunk 8]
```diff
@@ -35,15 +35,17 @@ func (e HTTPError) Error() string {
...
 }
 
-// TODO(xiangli): handle http write errors
-func (e HTTPError) WriteTo(w http.ResponseWriter) {
+func (e HTTPError) WriteTo(w http.ResponseWriter) error {
 	w.Header().Set("Content-Type", "application/json")
 	w.WriteHeader(e.Code)
 	b, err := json.Marshal(e)
 	if err != nil {
 		plog.Panicf("marshal HTTPError should never fail (%v)", err)
```
Comment (gyuho) [inline etcdserver/etcdhttp/httptypes/errors.go]: @xiang90 I put this back. Should I send another PR squashing all these changes?

Thanks,

Comment (xiang90) [inline etcdserver/etcdhttp/httptypes/errors.go]: you can squash all and force push onto the same branch.

Comment (gyuho) [inline etcdserver/etcdhttp/httptypes/errors.go]: Ok got it. I will. Thanks.

File: storage/watchable_store.go [hunk 9]
```diff
@@ -305,7 +301,7 @@ func (s *watchableStore) notify(rev int64, ev storagepb.Event) {
 					}
 				}
 				w.cur = rev
-				s.unsynced = append(s.unsynced, w)
+				s.unsynced[w] = struct{}{}
 			}
 		}
```
Comment (gyuho) [inline storage/watchable_store.go]: This change was not meant to be here. I will delete this change when squashing.

**Comment by gyuho** [comment]
@xiang90 just combined all previous commits into one.
And thanks for the review again!

**Comment by xiang90** [comment]
@gyuho Thanks for your contribution!

**Comment by gyuho** [comment]
Thank you!
